sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1279,8 +1409,26 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
       (c, evPrim, evNull) => code"$evNull = true;"
     case TimestampType =>
       (c, evPrim, evNull) => code"$evPrim = (long) ${timestampToIntegerCode(c)};"
+    case DecimalType() if failOnIntegerOverflow =>
+      (c, evPrim, evNull) =>
+        code"""
+          try {
+            $evPrim = $c.toJavaBigInteger().longValueExact();
+          } catch (ArithmeticException e) {
+            throw new ArithmeticException("Casting $c to Long causes overflow");
+          }
+        """
     case DecimalType() =>
       (c, evPrim, evNull) => code"$evPrim = $c.toLong();"
+    case _: FloatType | _: DoubleType if failOnIntegerOverflow =>
+      (c, evPrim, evNull) =>
+        code"""
+          if ($c >= ${Long.MinValue}L && $c <= ${Long.MaxValue}L) {
```

Comment:
Do we expect that this check matches the Nan case, too?
```
scala> sql("SET spark.sql.failOnIntegerOverFlow=true")
scala> sql("select cast(float('NaN') as long)").show
java.lang.ArithmeticException: Casting NaN to Long causes overflow.
  at org.apache.spark.sql.types.FloatExactNumeric$.toLong(numerics.scala:137)
```

Reply:
I am not sure about that. Keeping the current behavior seems fine as well, since we have printed the exact value which causes casting failure.

Reply:
If this is an expected behaivour, its ok to me. pgSQL handles the out-of-range case and NaN case in the same way: https://github.com/postgres/postgres/blob/927f34ce8a215c8b254136f710cca9ca4da1352c/src/backend/utils/adt/int8.c#L1269